src/api/v4/layer/layer.js

```diff
@@ -217,12 +233,27 @@ Layer.prototype.getFeatureClickColumns = function (columns) {
  * @api
  */
 Layer.prototype.setFeatureOverColumns = function (columns) {
-  this._featureOverColumns = columns;
-  if (this._internalModel) {
-    this._internalModel.tooltip.update(_getInteractivityFields(columns));
+  var prevColumns = this._featureOverColumns;
+  _checkColumns(columns);
+  if (prevColumns === columns) {
+    return Promise.resolve();
   }
-
-  return this;
+  // If layer is not instantiated just store the new status
+  if (!this._internalModel) {
+    this._featureOverColumns = columns;
+    return Promise.resolve();
+  }
+  // Update the internalModel and return a promise
+  this._internalModel.tooltip.fields.set(_getInteractivityFields(columns).fields, { silent: true });
+  return this._engine.reload()
+    .then(function () {
+      this._featureOverColumns = columns;
+    }.bind(this))
+    .catch(function (err) {
```

Comment:
Boy scout rule here:  I think this 3 lines could be extracted into a function to reduce LOC

```javascript
function _rejectAndTriggerError(err){
  var error = new CartoError(err);
  this.trigger(EVENTS.ERROR, error);
  return Promise.reject(error);
}
```

And use:

```javascript
.catch(_rejectAndTriggerError.bind(this));
```

Reply:
:+1: That pattern is used in every promise rejection in the file so we can refactor the four usages.